text/0000-generalized-arity-tuples.md

```diff
@@ -0,0 +1,277 @@
+- Although the intention is to make code more understandable, depending on how the documentation is rendered, it could become _less_ understandable.
+- Someone might find a better, more general or simpler solution after this RFC has been implemented. In this case, it will be hard to remove the current solution.
+
+# Rationale and alternatives
+[rationale-and-alternatives]: #rationale-and-alternatives
+
+The selling point of the proposed solution is that it is completely based on existing concepts. The syntax and type system remain unaffected. Hence, the implementation effort should be predictable and the risk of compromising the overall quality of the language should be low. A second benefit is the possibility to define more advanced type mappings, e.g. `(A, B, C, ..., Z)` &rarr; `(B, A, C, ..., Z)`.
+
+An alternative approach to the tuple generalization problem could be to add some kind of type-level iterator to the language. Although the idea seems simple and straight-forward at first sight, it comes with some big drawbacks:
+
+- New syntax for the iteration must be introduced. This, usually, is a very hard problem.
+- The compiler needs a new type-level iterator machinery.
+- It is hard to imagine how more advanced type mappings can be realized without introducing even more syntax.
+
+# Prior art
+[prior-art]: #prior-art
+
+Similar solutions have been proposed in earlier RFCs. The drawbacks compared to this RFC are summarized here:
+
+- [#1582](https://github.com/rust-lang/rfcs/pull/1582):
+  - Introduces new syntax
+  - Introduces new traits
+  - Deals with memory layout issues
+- [#1921](https://github.com/rust-lang/rfcs/pull/1921):
+  - Focussed on variadic function arguments
+  - Introduces new attributes
+  - Changes the way the language works by introducing function overloading
+- [#1935](https://github.com/rust-lang/rfcs/pull/1935)
+  - Focussed on variadic generics
+  - Introduces new syntax
+  - Includes new traits
+  - Uses special handling for references
+
+# Unresolved questions
+[unresolved-questions]: #unresolved-questions
+
+Those points are mainly decisions to be made before implementing the RFC:
+
+- How should compiler messages or the documentation be rendered? The printed output for `Tuple<A, Tuple<B, Tuple<C, ()>>>` must probably be mapped back to `(A, B, C)` for readability. But what if this reverse mapping is impossible as is the case for the generalized tuple `impl`s?
+- What should the compiler do with nonsensical tuples? A nonsensical tuple is a `Tuple` whose `TAIL` parameter is not a tuple (e.g. `Tuple<String, String>`). It feels like the easiest and most idiomatic answer is that the compiler should not care and let the code run into a type error as soon as the tuple is used. Nevertheless, nonsensical tuples could be discovered and reported by `clippy`.
```

Comment:
I agree with not doing anything about "nonsensical" tuples; seems like banning them just brings unjustified complication to the type system and undoes the nice simplification benefits your proposal brings.

Reply:
Rust's current type system is plenty enough to be able to specify that a tuple cons-element can only have `()` or another tuple cons-element as its tail associated type, so a separate mechanism for checking the well-formedness of a tuple type list is not needed. (Custom diagnostics might be of help in that area, though.)

Reply:
Can't we just have a closed trait `IsTuple` that is only implemented by valid tuples? That would be a simple modification, but would avoid any extra complexities by not making sure `Tuple` is well-formed?